File: leetcode-cn/4.寻找两个正序数组的中位数.py
#
# @lc app=leetcode.cn id=4 lang=python3
#
# [4] 寻找两个正序数组的中位数
#

# @lc code=start
from typing import List


class Solution:
    def findMedianSortedArrays(self, nums1: List[int],
                               nums2: List[int]) -> float:
        def help(nums1, start1, end1, nums2, start2, end2, k):
            len1 = end1 - start1 + 1
            len2 = end2 - start2 + 1

            # 保证 nums1 较短
            if len1 > len2:
                return help(nums2, start2, end2, nums1, start1, end1, k)

            if len1 == 0:
                return nums2[start2 + k - 1]

            if k == 1:
                return min(nums1[start1], nums2[start2])

            i = start1 + min(k // 2, len1) - 1
            j = start2 + min(k // 2, len2) - 1

            if nums1[i] <= nums2[j]:
                return help(nums1, i + 1, end1, nums2, start2, end2,
                            k - (i - start1 + 1))
            else:
                return help(nums1, start1, end1, nums2, j + 1, end2,
                            k - (j - start2 + 1))

        m = len(nums1)
        n = len(nums2)

        if (m + n) % 2 == 1:
            return help(nums1, 0, m - 1, nums2, 0, n - 1, (m + n + 1) // 2)
        else:
            return (help(nums1, 0, m - 1, nums2, 0, n - 1,
                         (m + n) // 2) + help(nums1, 0, m - 1, nums2, 0, n - 1,
                                              (m + n) // 2 + 1)) / 2


# @lc code=end

File: leetcode-cn/C++/4.寻找两个正序数组的中位数.cpp
/*
 * @lc app=leetcode.cn id=4 lang=cpp
 *
 * [4] 寻找两个正序数组的中位数
 */

// @lc code=start
#include <vector>

using namespace std;

class Solution
{
public:
    int help(vector<int> nums1, vector<int> nums2, int k)
    {
        int m = nums1.size();
        int n = nums2.size();

        int index1 = 0;
        int index2 = 0;

        while (true)
        {
            if (index1 == m)
            {
                return nums2[index2 + k - 1];
            }
            if (index2 == n)
            {
                return nums1[index1 + k - 1];
            }
            if (k == 1)
            {
                return min(nums1[index1], nums2[index2]);
            }

            int new_index1 = min(index1 + k / 2 - 1, m - 1);
            int new_index2 = min(index2 + k / 2 - 1, n - 1);

            if (nums1[new_index1] <= nums2[new_index2])
            {
                k -= (new_index1 - index1 + 1);
                index1 = new_index1 + 1;
            }
            else
            {
                k -= (new_index2 - index2 + 1);
                index2 = new_index2 + 1;
            }
        }
    }

    double findMedianSortedArrays(vector<int> &nums1, vector<int> &nums2)
    {
        int m = nums1.size();
        int n = nums2.size();

        if ((m + n) % 2 == 1)
        {
            return help(nums1, nums2, (m + n + 1) / 2);
        }
        else
        {
            return (help(nums1, nums2, (m + n) / 2) + help(nums1, nums2, (m + n) / 2 + 1)) / 2.0;
        }
    }
};
// @lc code=end
